### ISSUE TITLE
[Issue]: Upgrade GHA usages of `actions/upload-artifact` and `actions/download-artifact` in CI/CD pipelines

### ISSUE THREAD

**Issue by TylerHendrickson** [usdigitalresponse/usdr-gost#2528]
### Why is this issue important?

This repository is currently using v3.x of the `actions/upload-artifact` and `actions/download-artifact` GitHub actions. In addition to generally preferring up-to-date dependencies, the fact that these actions use the Node.js 16 runtime contributes to noise in our GitHub Actions workflow annotations, with output like this:
> Node.js 16 actions are deprecated. Please update the following actions to use Node.js 20: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32. For more information see: https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/.

Additionally, upgrading to v4.x should yield the following benefits:
- > Downloads are significantly faster, upwards of 90% improvement in worst case scenarios.
  https://github.com/actions/download-artifact?tab=readme-ov-file#improvements
- > Using upload-artifact@v4 ensures artifacts are immutable, improving performance and protecting objects from corruption, which would often happen with concurrent uploads.
  https://github.blog/changelog/2023-12-14-github-actions-artifacts-v4-is-now-generally-available/

### Current State

The following workflow files contain jobs that use v3.x of `actions/upload-artifact` and/or `actions/download-artifact`:
- `deploy-production.yml`
- `build.yml`
- `terraform-apply.yml`
- `terraform-plan.yml`

### Expected State

All workflows in this repository use the latest 4.x version of `actions/upload-artifact` and `actions/download-artifact`, with versions pinned to the immutable SHA for that version, e.g.
```yaml
uses: actions/download-artifact@6b208ae046db98c579e8a3aa621ab581ff575935  # v4.1.1
```
```yaml
uses: actions/upload-artifact@26f96dfa697d77e81fd5907df203aa23a56210a8  # v4.3.0
```

### Implementation Plan

- Upgrade the pinned version of all usages of [`actions/upload-artifact`](https://github.com/actions/upload-artifact) to the latest version
- Address [breaking changes](https://github.com/actions/upload-artifact?tab=readme-ov-file#breaking-changes) for `actions/upload-artifact` (see [migration guide](https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md))
- Upgrade the pinned version of all usages of [`actions/download-artifact`](https://github.com/actions/download-artifact) to the latest version
- Address [breaking changes](https://github.com/actions/download-artifact?tab=readme-ov-file#breaking-changes) (see [migration guide](https://github.com/actions/download-artifact/blob/main/docs/MIGRATION.md))

**Comment by jeffsmohan** [comment]
I can take a look at this!

**Comment by TylerHendrickson** [comment]
Fixed in #2539
